org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/resource/Storage2UriMapperJavaImpl.java

```diff
@@ -352,49 +353,60 @@ protected Void handle(URI uri, IStorage storage,
+		 * - uriPrefix == null || uri startsWith uriPrefix
+		 *   => regularMatch
+		 * 
+		 * - uri.isArchive && archiveURI.isFile || isPlatformResource && uriPrefix is not null && no other matches && 
+		 *   => archiveMatch
+		 */
+		List<Pair<IStorage, IProject>> regularMatches = new ArrayList<Pair<IStorage,IProject>>(1);
+		List<Pair<IStorage, IProject>> archiveMatches = null;
+		IPath archivePath = null;
+		if (uri.isArchive()) {
+			String authority = uri.authority();
+			URI archiveURI = URI.createURI(authority.substring(0, authority.length() - 1));
+			if (archiveURI.isFile() || archiveURI.isPlatformResource()) {
+				archivePath = new Path(archiveURI.isPlatformResource()? archiveURI.toPlatformString(true): archiveURI.toFileString());
+				archiveMatches = new ArrayList<Pair<IStorage,IProject>>(1);
+			}
+		}
+		for(PackageFragmentRootData data: packageFragmentRootDatas) {
+			if (data.uriPrefix == null || uri.toString().startsWith(data.uriPrefix.toString())) {
+				IStorage storage = data.uri2Storage.get(uri);
+				if (storage != null && data.exists()) {
+					synchronized(data.associatedRoots) {
 						for (IPackageFragmentRoot root : data.associatedRoots.values()) {
-							result.add(Tuples.create(storage, root.getJavaProject().getProject()));
+							regularMatches.add(Tuples.create(storage, root.getJavaProject().getProject()));
+							archiveMatches = null;
 						}
 					}
 				}
-			} else {
-				iterator.remove();
-			}
-		}
-		if (result.isEmpty() && uri.isArchive()) {
-			String authority = uri.authority();
-			authority = authority.substring(0, authority.length() - 1);
-			URI archiveURI = URI.createURI(authority);
-			if (archiveURI.isFile() || archiveURI.isPlatformResource()) {
-				IPath archivePath = new Path(archiveURI.isPlatformResource()? archiveURI.toPlatformString(true): archiveURI.toFileString());
-				for (PackageFragmentRootData data : packageFragmentRootDatas) {
-					if (data.uriPrefix != null && archivePath.equals(data.getPath())) {
-						// prefixes have an empty last segment.
-						URI prefix = data.uriPrefix.lastSegment().length()==0 ? data.uriPrefix.trimSegments(1) : data.uriPrefix;
-						URI expectedURI = prefix.appendSegments(uri.segments());
-						IStorage storage = data.uri2Storage.get(expectedURI);
-						if (storage != null) {
-							for (IPackageFragmentRoot root : data.associatedRoots.values()) {
-								result.add(Tuples.create(storage, root.getJavaProject().getProject()));
-							}
+			} else if (archiveMatches != null && archivePath != null && data.uriPrefix != null && archivePath.equals(data.getPath())) {
+				// prefixes have an empty last segment.
+				URI prefix = data.uriPrefix.lastSegment().length()==0 ? data.uriPrefix.trimSegments(1) : data.uriPrefix;
+				URI expectedURI = prefix.appendSegments(uri.segments());
+				IStorage storage = data.uri2Storage.get(expectedURI);
+				if (storage != null && data.exists()) {
+					synchronized (data.associatedRoots) {
```

Comment:
other placeds dont synchronize on associatedRoots

Reply:
It's a synchronized collection. Atomic put / get don't need to synchronize. Here we iterate so I figured it might be best to sync.

Reply:
so the getFirst at other places wont harm?

Reply:
Good catch. I didn't think about this being two calls. I'll refactor this.

Reply:
Done.